### ISSUE TITLE
Buttons and links should have meaningful content

### ISSUE THREAD

**Issue by alvaromontoro** [jackdomleo7/Checka11y.css#34]
### Describe the new a11y feature or project enhancement

Buttons and links should contain something that indicates what they are or what they are going to go. Having an empty link, or a button that is only an image, or an empty input button without a value can be confusing and doesn't provide enough information for screen reader users.

From [WCAG 2.1 Success Criterion 2.4.4 Link Purpose (In Context)](https://www.w3.org/TR/WCAG21/#link-purpose-in-context): "_The purpose of each link can be determined from the link text alone or from the link text together with its programmatically determined link context [...]_."

### Describe the solution you'd like

Verify that buttons and links follow some basic patterns:

- They are not empty (e.g. using the `:empty` pseudo-class)
- They don't contain just an image tag (e.g. using the `:only-child` pseudo-class)

### Link(s)

- [WCAG 2.1: Success Criterion 2.4.4 Link Purpose (In Context)](https://www.w3.org/TR/WCAG21/#link-purpose-in-context)
- [WebAIM: Links and hypertext](https://webaim.org/techniques/hypertext/link_text#empty_links)

**Comment by jackdomleo7** [comment]
Hmm.... I like this idea. I'd be interested to the solution. 🤔

I'm curious because the below code:

```html
<button aria-label="View shopping cart">
  <svg .....></svg>
</button>
```

I just as accessible as (providing `.sr-only` has the correct styles to hide the content):

```html
<button>
  <span class="sr-only">View shopping cart</span>
  <svg .....></svg>
</button>
```

I personally prefer the `aria-label` opion because it then allows you to accessibly apply CSS-only tooltips such as [Cooltipz.css](https://cooltipz.jackdomleo.dev), so a non-screen-reader user can read it.

**Comment by danspratling** [comment]
If both are valid, to avoid _opinionization_ why don't we just check at least one of those two approaches exist?

**Comment by jackdomleo7** [comment]
I agree, I wanted to clarify what @alvaromontoro was saying. 🙂

**Comment by danspratling** [comment]
I think we have enough context for this to get picked up.

- [ ] buttons are not empty
- [ ] buttons which contain only an image are aria-labelled
- [ ] buttons which contain only an image and no aria-label error
- [ ] links (`<a>`) are not empty
- [ ] links (`<a>`) which contain only an image are aria-labelled
- [ ] links (`<a>`) which contain only an image and no aria-label error

**Comment by alvaromontoro** [comment]
@jackdomleo7 The `:not()` modifier could be applied so only the buttons that don't have an aria-label are targeted. Something like this:

```css
button:not([aria-label]):not([aria-labelledby]),
button:not([aria-label]):not([aria-labelledby]) img:only-child,
button:not([aria-label]):not([aria-labelledby]) svg:only-child {
  {error}
}
```

But that would have some false positives, as you can see on this demo: https://codepen.io/alvaromontoro/pen/vYGqxBN

**Comment by jackdomleo7** [comment]
Hmm ok. Good demo by the way. What do you think is the best solution?

Ideally, what we want is a feature that doesn't throw any false errors - so if that means not considering every option for now, then I'm happy with that. Part of a feature is better than no feature. 🙂

**Comment by alvaromontoro** [comment]
As far as I know, there is no way to check if there's a text node alongside the element. So to avoid false positives, remove the check for image as only child, and check for a button that is empty and does not have `aria-label` or `aria-labelledby`.

**Comment by jackdomleo7** [comment]
Sounds good to me! I'd love to see a PR with various examples. 🙂

**Comment by alvaromontoro** [comment]
This can be extended to `label` too.

**Comment by tricinel** [comment]
@jackdomleo7 if it's alright with everyone, I'll give it a shot.
